### ISSUE TITLE
How to programmatically quit Voilà?

### ISSUE THREAD

**Issue by enricogandini** [voila-dashboards/voila#864]
Thanks everybody for the awesome Voilà project!

I wanted to ask if there is a way to programmatically quit Voilà, killing the Python kernel.
I am building a Voilà app based on [asynchronous ipywidgets](https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20Asynchronous.html). I would like to quit the app and kill the Python kernel when the user performs some action on the widgets.

I tried `quit()`, `exit()`, and `sys.exit()`, and I executed Voilà with `--debug` option. All the aforementioned functions make Voilà print a message on the terminal: `[Voila] AsyncIOLoopKernelRestarter: restarting kernel (1/5), keep random ports`. It looks like the aforementioned functions try to kill the Python kernel, which is immediately restarted by Voilà.

Is there some other function that I should be using to kill Voilà and the associated kernel? Or is there some Voilà option that I am not aware of?

**Comment by agoose77** [comment]
I don't know whether Voila does this, but you can implement this yourself by invoking `os.kill` with `os.getppid`, if you have the right permissions: https://docs.python.org/3/library/os.html#os.getppid

**Comment by enricogandini** [comment]
> I don't know whether Voila does this, but you can implement this yourself by invoking `os.kill` with `os.getppid`, if you have the right permissions: https://docs.python.org/3/library/os.html#os.getppid

Thanks!
I tried using `os.kill(os.getppid())`, but the terminal on which I launched Voilà is still using Voilà. When i launch Voilà with option `--debug`, from the terminal I get the message `[Voila] activity on 038cadf3-7fcc-4d6b-aeb1-c0af1f4c76ee: error`.
The browser tab on which the Voilà app was opened is still open.

**Comment by agoose77** [comment]
What OS are you using?

**Comment by martinRenou** [comment]
`os.kill(os.getppid())` would only kill the kernel, not the voila application. 

But you might be able to do it if you know the Voila pid?

**Comment by enricogandini** [comment]
I am using Ubuntu 18.04; I think I have all the necessary permissions, since I am still testing the app on my local machine. When I will deploy it, I do not know if I will still have all necessary permissions. But I am really interested in knowing what would be the best way to kill the Voilà application.

**Comment by martinRenou** [comment]
Maybe something like this would work:
```python
import psutil

pid = None

for proc in psutil.process_iter():
    if 'voila' in proc.name():
       os.kill(proc.pid)
```

Please do not hold me responsible if anything goes wrong though 😄

**Comment by agoose77** [comment]
~~@martinRenou shouldn't `getppid` return the PID of the parent process?~~ ah, it seems that this only applies to `fork`

**Comment by agoose77** [comment]
Another option is to use a wrapper to launch voila, and capture the PID
```bash
#!/usr/bin/env bash
set -eu
export CAPTURED_PID="$$"
exec "$@"
```
`./capture voila ...` would launch voila with the PID stored in `CAPTURED_PID`. I assume that kernels inherit all the appropriate ENV vars, and that killing this process would kill the child kernel processes.

**Comment by enricogandini** [comment]
Thanks for your answers @martinRenou and @agoose77!
I tried both your approaches, but I keep getting the same error when running Voilà with `--debug` option, and neither the browser page nor the Voilà process on the terminal are terminated.

**Comment by martinRenou** [comment]
> neither the browser page

The page will not close itself if you kill the voila process. The user will just get a non-responsive page, and they might not even understand what happened.

**Comment by enricogandini** [comment]
Yes, the page is in fact non-responsive after I try to kill the process with any of the aforementioned methods.
But the terminal is still running Voilà, and the user can still reload the page to restart the app. I see the error message, but Voilà is not really killed.

Maybe I am asking for something impossible or unreasonable, I am not an expert in web development.

**Comment by agoose77** [comment]
@enricogandini it's probably an error in your script. This works with my capture wrapper:
```json
{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": ["import os,signal;os.kill(int(os.environ['CAPTURED_PID']),signal.SIGKILL)"]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
```

**Comment by enricogandini** [comment]
Yes, I did not convert the `CAPTURED_PID` from string to int, and I did not include `signal.SIGKILL`.
Executing `voila` through the `capture` wrapper described by @agoose77, and including `os.kill(int(os.environ["CAPTURED_PID"]), signal.SIGKILL)` in my notebook, produces an unresponsive browser tab (that cannot be reloaded by the user), and kills the Voilà process in the terminal.

Thanks a lot for your answers!

**Comment by agoose77** [comment]
@enricogandini I tested with `getppid` and it works for me too, i.e. 

```python
import os, signal
os.kill(os.getppid(), signal.SIGKILL)
```

**Comment by agoose77** [comment]
To give the user a more friendly UX, you would want to use JS to trigger the browser tab to close.

**Comment by enricogandini** [comment]
Yes, using `os.getppid` works just as well, and is more convenient than the `capture` wrapper. It is necessary to include `signal.SIGKILL`.

I am sure that using JavaScript to kill the process would give users a better experience, but I don't know anything about JavaScript, I was looking for a Python solution, and I am satisfied with it!

**Comment by agoose77** [comment]
RE using JS, I mean that you could use it alongside the shutdown logic. The best way would probably be to add some JS that connects to the kernel, and triggers the window close when it detects that the connection is lost.

**Comment by enricogandini** [comment]
Thanks for the suggestion, I will find out how to execute JS code inside a Python Jupyter notebook.

But I am quite satisfied with the Python solution you provided. As I mentioned at the beginning of this issue, I am writing an app which uses asynchronous widgets. So, when the user reaches reaches the end of the app (after passing through many widgets), I show him one last "Thank you, goodbye!" widget, and I kill the process so he cannot reload the page. In my opinion, this is a reasonably good user experience!

**Comment by chuckmandu** [comment]
> RE using JS, I mean that you could use it alongside the shutdown logic. The best way would probably be to add some JS that connects to the kernel, and triggers the window close when it detects that the connection is lost.

How to detect IPython kernel events in Javascript within the Voila environment is where I get stuck. Any guidance is appreciated.

**Comment by RexBarker** [comment]
I used this script to run Voila and shutdown when it determines the kernel has stopped.  Additional arguments to Voila in the `cmd` statement:

```python
# Wrapper script to stop the Voila server after kernel has terminated
import subprocess
import sys


def run_voila(notebook_path):
    cmd = [
        "voila.exe",
        "--theme=dark",
        '--VoilaConfiguration.file_whitelist=[".*"]',
        notebook_path,
    ]

    with subprocess.Popen(
        cmd,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        bufsize=1,
        universal_newlines=True,
        text=True,
    ) as process:
        try:
            for line in process.stdout:
                print(line, end="")  # Real-time output
                if "Kernel shutdown" in line:
                    print("\nDetected kernel shutdown - terminating Voilà")
                    process.terminate()
                    sys.exit(0)
        except KeyboardInterrupt:
            process.terminate()
            sys.exit(1)


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python script.py <notebook-path>")
        sys.exit(1)

    run_voila(sys.argv[1])
```
